pkg/gatewayserver/io/ws/lbslns/upstream.go

```diff
@@ -214,25 +212,19 @@ func (req *JoinRequest) toUplinkMessage(ids ttnpb.GatewayIdentifiers, bandID str
 		return nil, errJoinRequestMessage.WithCause(err)
 	}
 
-	timestamp := uint32(req.RadioMetaData.UpInfo.XTime & 0xFFFFFFFF)
-
-	var rxTime *time.Time
-	sec, nsec := math.Modf(req.RadioMetaData.UpInfo.RxTime)
-	if sec != 0 {
-		val := time.Unix(int64(sec), int64(nsec*(1e9)))
-		rxTime = &val
-	}
-
-	rxMetadata := &ttnpb.RxMetadata{
-		GatewayIdentifiers: ids,
-		Time:               rxTime,
-		Timestamp:          timestamp,
-		Rssi:               req.RadioMetaData.UpInfo.RSSI,
-		ChannelRssi:        req.RadioMetaData.UpInfo.RSSI,
-		Snr:                req.RadioMetaData.UpInfo.SNR,
-		AntennaIndex:       uint32(req.RadioMetaData.UpInfo.RCtx),
+	timestamp := TimestampFromXTime(req.RadioMetaData.UpInfo.XTime)
+	tm := TimePtrFromGPSTime(req.UpInfo.GPSTime)
+	up.RxMetadata = []*ttnpb.RxMetadata{
+		{
+			GatewayIdentifiers: ids,
+			Time:               tm,
```

Comment:
Note that `up.Settings.Time` and `up.RxMetadata[0].Time` are now always `nil` if no `gpstime` is available, instead of being translations of `RxTime`. 

The reason is that `io.HandleUp` sees `up.Settings.Time` as a source of gateway GPS time, which is incorrect for `RxTime` - `RxTime` is not GPS-sourced, but rather the local system time at receive.

You may argue that 'it's almost as good' (since indeed, it has microsecond precision), but the following issue arises:

The presence of `up.Settings.Time` causes the GS to consider that it has successfully synchronized with gateway absolute time. This would allow something like a class B downlink to take place. The problem is that the transmission acknowledgement (the `dntxed`) does _not_ contain a `RxTime` - only a `gpstime`. This causes the GS to consider once again that the gateway absolute time synchronization has been lost, until the next uplink arrives.